### PR TITLE
wgsl: f16 execution test for binary matrix expression and built-in dot

### PR DESCRIPTION
**Pull request by jzm-intel**
This PR add execution tests for f16 binary matrix expression, including matrix addition/subtraction and matrix-scalar/vector/matrix multiplication, and also the execution test for 16 built-in dot.

Issue: #1248, #1626, #2513

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Comment by jzm-intel** [comment]
@zoddicus Hi, please take a look, thanks!

This PR has some overlap with #2951 as `transpose` op is needed. I would do the rebase if your PR is landed.

File: src/webgpu/shader/execution/expression/binary/f16_matrix_addition.spec.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import { binary, compoundBinary } from './binary.js';
+
+export const g = makeTestGroup(GPUTest);
+
+// Cases: matCxR_[non_]const
+const mat_cases = ([2, 3, 4] as const)
+  .flatMap(cols =>
+    ([2, 3, 4] as const).flatMap(rows =>
+      ([true, false] as const).map(nonConst => ({
+        [`mat${cols}x${rows}_${nonConst ? 'non_const' : 'const'}`]: () => {
```
Comment (jzm-intel) [inline src/webgpu/shader/execution/expression/binary/f16_matrix_addition.spec.ts:22]: Using the map-reduce method to generate testcases in order to significantly shorten the code and get rid of repeat code.

Comment (zoddicus) [reply src/webgpu/shader/execution/expression/binary/f16_matrix_addition.spec.ts:22]: Nice :+1:

Comment (zoddicus) [reply src/webgpu/shader/execution/expression/binary/f16_matrix_addition.spec.ts:22]: Filed https://github.com/gpuweb/cts/issues/2957 to track updating the existing tests to use this style of implementation

**Review by zoddicus — APPROVED** [review]

**Comment by zoddicus** [comment]
> @zoddicus Hi, please take a look, thanks!
> 
> This PR has some overlap with #2951 as `transpose` op is needed. I would do the rebase if your PR is landed.

f16 transpose is landed

**Comment by jzm-intel** [comment]
Thanks for reviewing! I am merging this PR.

Some matrix-vertex/matrix multiplication unittest cases are modified to ensure that result values are exactly representable in f16.

**Pull request merged**